exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/columnreaders/AsyncPageReader.java

```diff
@@ -297,11 +297,11 @@ private void waitForExecutionResult() throws InterruptedException, ExecutionExce
     while (asyncPageRead != null && !asyncPageRead.isEmpty()) {
       try {
         Future<Void> f = asyncPageRead.poll();
-        if(!f.isDone() && !f.isCancelled()){
+        if(!f.isDone() && !f.isCancelled()) {
           f.cancel(true);
-        } else {
-          f.get(1, TimeUnit.MILLISECONDS);
         }
+        // The framework guarantees a blocking version of FutureTask cancellation. At this time we are
```

Comment:
Please see `PartitionerTask` in `PartitionerDecorator`. The design goal for the `PartionerTask` is exactly the same, make sure that the main thread will not proceed till all tasks are either done or canceled. Currently, `PartitionerTask` is a private class, but it should be possible to generalize it.

Reply:
@vrozov 

I have looked at that fix but decided against it as I felt the same logic could be achieved with a simpler implementation especially that the performance penalty is only incurred during the cancellation process.

Reply:
My $0.02  is to tend towards  what @vrozov is suggesting. It would be preferable to have one reusable implementation that works for both use cases. 
I don't know if @sachouche needs the localStats in PartitionTask, nor, as he mentioned, do you want to pay a price for non cancellation paths. 
Is it possible to come up with a unified approach here?